main/main.c

```diff
@@ -19,18 +19,69 @@
 #include <stdio.h>
 #include <stdbool.h>
 #include <signal.h>
+#include <sys/time.h>
+#include <inttypes.h>
 
 #include <rte_eal.h>
 #include <rte_log.h>
 #include <rte_common.h>
 #include <rte_launch.h>
+#include <rte_cycles.h>
 
 #include "gatekeeper_config.h"
 #include "gatekeeper_net.h"
 
 /* Indicates whether the program needs to exit or not. */
 volatile int exiting = false;
 
+/*
+ * These metrics are system dependent, and 
+ * initialized via time_resolution_init() function.
+ */
+uint64_t cycles_per_sec = 0UL;
+uint64_t picosec_per_cycle = 0UL;
+
+/* Obtain the system time resolution. */
+static int
+time_resolution_init(void)
+{
+	int ret;
+	uint64_t tsc_start;
+	struct timeval tv_start;
+	struct timeval tv_now;
+
+	tsc_start = rte_rdtsc();
+	ret = gettimeofday(&tv_start, NULL);
+	if (ret < 0)
+		return ret;
+
+	while (1) {
+		uint64_t diff;
+		uint64_t tsc_now;
+
+		ret = gettimeofday(&tv_now, NULL);
+		if (ret < 0)
+			return ret;
+
+		diff = (uint64_t)(tv_now.tv_sec - tv_start.tv_sec) * 1000000UL +
+			(uint64_t)(tv_now.tv_usec - tv_start.tv_usec);
+
+		if (diff >= 1000000UL) {
+			tsc_now = rte_rdtsc();
+			cycles_per_sec = (tsc_now - tsc_start) * 1000000UL / diff;
+			break;
+		}
+	}
+
+	picosec_per_cycle = (1000000000000ULL / cycles_per_sec);
```

Comment:
`cycles_per_sec` is obtained from an integer division and picoseconds is way more precise than seconds, so obtaining `picosec_per_cycle` from the original parameters will improve precision.